### PR TITLE
Replace all uses of AtomicStateMachine and remove it

### PR DESCRIPTION
**Pull request by wkaluza**

### CODE REVIEW

File: libs/ledger/include/ledger/storage_unit/transaction_store_sync_service.hpp [hunk 1]
```diff
@@ -110,15 +109,21 @@ class TransactionStoreSyncService : public network::AtomicStateMachine<tx_sync::
...
   // We need this for the testing.
   bool IsReady()
   {
     FETCH_LOCK(is_ready_mutex_);
     return is_ready_;
   }
 
+  void Execute()
+  {
+    if (state_machine_->IsReadyToExecute())
```
Comment (pbukva) [inline libs/ledger/include/ledger/storage_unit/transaction_store_sync_service.hpp:122]: Just minor comment - in theory this creates potential for "hot" loop if ```state_machine_->IsReadyToExecute()``` returns ```false```.

To mitigate this potential risk, I would suggest to add escape valve in form of ```else{std::this_thread::sleep_for(...);}```

Comment (wkaluza) [reply libs/ledger/include/ledger/storage_unit/transaction_store_sync_service.hpp:122]: There is already a wait in this execution path - the lane service calls Execute using a "BackgroundedWorkThread" which locks on a condition variable for SYNC_PERIOD_MS (currently set to 500ms) between executions. See void LaneService::Start() in lane_service.cpp

File: libs/ledger/src/protocols/executor_rpc_client.cpp [hunk 2]
```diff
@@ -64,14 +71,17 @@ class ExecutorConnectorWorker : public network::AtomicStateMachine<ExecutorRpcCl
...
     }
     catch (...)
     {
       FETCH_LOG_WARN(LOGGING_NAME, "Work threw.");
       throw;
     }
-    auto r = this->AtomicStateMachine::Get();
+    auto r = state_machine_->state();
 
     switch (r)
```
Comment (pbukva) [inline libs/ledger/src/protocols/executor_rpc_client.cpp:85]: Current Implementation is just mimicking 1:1 all states of completely unrelated workflow of lower layer (`Promise` class).
This class (higher layer) should define its own sets of states (and related workflow).
I know that it was like that before the refactoring, but I would suggest rework this now.

The implication of this issue, is that, for example the `State::FAILED` is never triggered at all, it just sits here with related callback method as dead code.

I would suggest to clearly state states & workflow which suits this state machine, for example something like:
```
  enum class State
  {
    INITIAL = 0,
    CONNECTING,
    CONNECTED,
    TIMEDOUT
  };
```

NOTE:
Another thing here is (but it is **out** of scope of this PR), that **timeout** scenario shall be handled by lower layer(s) `Promise` (potentially by thread worker class) by exposing API methods to allow to set timeout, which would be then automatically evaluated by these lower layers, where such API would of course offer way for upper layer to be notified when set timeout actually times out (e.g. callback, or getter method, etc. ...).
And similarly, handling of other common states INITIAL, FAILED and FINISHED scenarios could be also relocated to lower layers.

**Comment by engineerkatie** [comment]
Noooooooooooooooo!!!!!!!!!

AtomicStateMachine is the coolest named C++ component ever in the history
of EVER!!!

File: libs/ledger/src/protocols/executor_rpc_client.cpp (hunk 2 above)
Comment (wkaluza) [reply libs/ledger/src/protocols/executor_rpc_client.cpp:85]: The issue with FAILED state came up this morning - it's a broader problem with the ExecutorRpcClient, in that it isn't code that currently gets used much outside of its own tests. According to @ejfitzgerald, once the time comes to pry Constellation apart, this class will need a rethink, but it's not a priority right now.

File: libs/ledger/include/ledger/storage_unit/transaction_store_sync_service.hpp (hunk 1 above)
Comment (pbukva) [reply libs/ledger/include/ledger/storage_unit/transaction_store_sync_service.hpp:122]: Keep in mind, that worker thread type is injected as template parameter (currently set to `BackgroundedWorkThread`), and that  worker thread type implements `Wait(...)` & `Wake(...)` methods, which can be implemented by other Type (if used in template instead) as empty methods (not really doing anything).

File: libs/ledger/src/protocols/executor_rpc_client.cpp (hunk 2 above)
Comment (pbukva) [reply libs/ledger/src/protocols/executor_rpc_client.cpp:85]: What I'm saying here is that you can (and should) drop failed from `ExecutorRpcClient::State` enum for reasons stated at the top of my comment.

The other things I wrote in NOTE, I clearly said that that is out of scope of this PR.

Comment (wkaluza) [reply libs/ledger/src/protocols/executor_rpc_client.cpp:85]: That's fair enough.

File: libs/ledger/include/ledger/storage_unit/transaction_store_sync_service.hpp (hunk 1 above)
Comment (wkaluza) [reply libs/ledger/include/ledger/storage_unit/transaction_store_sync_service.hpp:122]: I see your point, but I don't see how we could prevent future mistakes like that. I think it's fair enough, that when someone changes the worker thread to one with different behaviour, the resulting system will behave differently. Hopefully we'll be guarded against surprises like that as we build up our test coverage.

File: libs/ledger/include/ledger/protocols/executor_rpc_client.hpp [hunk 3]
```diff
@@ -56,11 +55,10 @@ class ExecutorRpcClient : public ExecutorInterface
   std::shared_ptr<Client> client;
 
   explicit ExecutorRpcClient(NetworkManager const &tm, Muddle &muddle)
-    : network_manager_(tm)
-  {
-    client_ = std::make_shared<Client>("R:Exec", muddle.AsEndpoint(), Muddle::Address(),
-                                       SERVICE_EXECUTOR, CHANNEL_RPC);
-  }
+    : client_(std::make_shared<Client>("R:Exec", muddle.AsEndpoint(), Muddle::Address(),
```
Comment (n-hutton) [inline libs/ledger/include/ledger/protocols/executor_rpc_client.hpp:58]: Unrelated to you: this doesn't need to be explicit. Also I think rename tm to nm here.

I suspect the order here should be the network manager is created before the client_, but perhaps it doesn't matter. It looks like network_manager_ is unused actually?

File: libs/ledger/include/ledger/storage_unit/transaction_store_sync_service.hpp [hunk 4]
```diff
@@ -51,8 +51,7 @@ enum class State
 };
 }
 
-class TransactionStoreSyncService : public network::AtomicStateMachine<tx_sync::State>,
-                                    public VerifiedTransactionSink
+class TransactionStoreSyncService : public VerifiedTransactionSink
```
Comment (n-hutton) [inline libs/ledger/include/ledger/storage_unit/transaction_store_sync_service.hpp:54]: 👍 👍 👍

File: libs/ledger/include/ledger/storage_unit/transaction_store_sync_service.hpp [hunk 5]
```diff
@@ -140,11 +145,21 @@ class TransactionStoreSyncService : public network::AtomicStateMachine<tx_sync::
...
+  State OnInitial();
+  State OnQueryObjectCounts();
+  State OnResolvingObjectCounts();
+  State OnQuerySubtree();
+  State OnResolvingSubtree();
+  State OnQueryObjects();
+  State OnResolvingObjects();
+  State OnTrimCache();
+
+  std::shared_ptr<core::StateMachine<State>> state_machine_;
```
Comment (n-hutton) [inline libs/ledger/include/ledger/storage_unit/transaction_store_sync_service.hpp]: normally style is using StateMachine

File: libs/ledger/src/protocols/executor_rpc_client.cpp [hunk 6]
```diff
@@ -64,78 +70,72 @@ class ExecutorConnectorWorker : public network::AtomicStateMachine<ExecutorRpcCl
   {
     try
     {
-      this->AtomicStateMachine::Work();
+      if (state_machine_->IsReadyToExecute())
```
Comment (n-hutton) [inline libs/ledger/src/protocols/executor_rpc_client.cpp:73]: I am not particularly familiar with the executor_rpc_client stuff. This seems like the state machine isn't following the usual reactor pattern and is instead being driven here - is that correct?

File: libs/ledger/src/storage_unit/lane_service.cpp [hunk 7]
```diff
@@ -171,9 +171,10 @@ void LaneService::Start()
 
   tx_sync_service_->Start();
 
-  // TX Sync service
-  workthread_ = std::make_shared<BackgroundedWorkThread>(
-      &bg_work_, "BW:LS-" + std::to_string(cfg_.lane_id), [this]() { tx_sync_service_->Work(); });
+  // TX Sync service - attach to reactor once #892 is merged
```
Comment (n-hutton) [inline libs/ledger/src/storage_unit/lane_service.cpp:174]: Ah, this sort of answers comment above

File: libs/ledger/src/storage_unit/transaction_store_sync_service.cpp [hunk 8]
```diff
@@ -64,25 +66,36 @@ const std::size_t TransactionStoreSyncService::BATCH_SIZE = 30;
...
+                                  &TransactionStoreSyncService::OnResolvingSubtree);
+  state_machine_->RegisterHandler(State::QUERY_OBJECTS, this,
+                                  &TransactionStoreSyncService::OnQueryObjects);
+  state_machine_->RegisterHandler(State::RESOLVING_OBJECTS, this,
+                                  &TransactionStoreSyncService::OnResolvingObjects);
+  state_machine_->RegisterHandler(State::TRIM_CACHE, this,
+                                  &TransactionStoreSyncService::OnTrimCache);
+
+  state_machine_->OnStateChange([](State new_state, State /* old_state */) {
+    (void)new_state;
```
Comment (n-hutton) [inline libs/ledger/src/storage_unit/transaction_store_sync_service.cpp]: There is a macro for this: FETCH_UNUSED(x)

File: libs/network/include/network/generics/atomic_state_machine.hpp [hunk 9]
```diff
@@ -1,154 +0,0 @@
-#pragma once
```
Comment (n-hutton) [inline libs/network/include/network/generics/atomic_state_machine.hpp:1]: 💯

File: libs/ledger/src/storage_unit/transaction_store_sync_service.cpp [hunk 10]
```diff
@@ -95,272 +108,255 @@ TransactionStoreSyncService::~TransactionStoreSyncService()
...
+  }
+
+  if (counts.failed)
+  {
+    FETCH_LOG_WARN(LOGGING_NAME, "Lane ", cfg_.lane_id, ": ", "Failed promises: ", counts.failed);
   }
-  };
 
-  if (result)
+  return State::TRIM_CACHE;
```
Comment (n-hutton) [inline libs/ledger/src/storage_unit/transaction_store_sync_service.cpp:349]: Looks good, the github diff tool has broken at the lots of changes here so I'll maybe test your branch over the weekend, to see if the subtree sync still functions (not sure there is a unit test for this atm).

**Review by n-hutton — CHANGES_REQUESTED** [review]
See last comment

File: libs/ledger/include/ledger/storage_unit/transaction_store_sync_service.hpp (hunk 1 above)
Comment (n-hutton) [reply libs/ledger/include/ledger/storage_unit/transaction_store_sync_service.hpp:122]: I wonder if this BackgroundedWorkThread is 'scheduled for demolition' as Ed would put it

Comment (n-hutton) [reply libs/ledger/include/ledger/storage_unit/transaction_store_sync_service.hpp:122]: Ed: ya

File: libs/ledger/include/ledger/protocols/executor_rpc_client.hpp (hunk 3 above)
Comment (wkaluza) [reply libs/ledger/include/ledger/protocols/executor_rpc_client.hpp:58]: Good spot. I'll remove the NetworkManager member and leave the constructor explicit.

File: libs/ledger/src/storage_unit/transaction_store_sync_service.cpp (hunk 10 above)
Comment (wkaluza) [reply libs/ledger/src/storage_unit/transaction_store_sync_service.cpp:349]: I'll look into writing some tests if they're not in place already.

**Review by n-hutton — APPROVED** [review]
LGTM

**Pull request merged**